Paginated view with updated domain names

sineed opened the pull request:
## The problem
I had a request to add pagination to Expedia channel app. I thought it will be as easy as for b.com but I was totally wrong. The problem was that b.com uses different entity names:
- RemoteAccount -> RemoteRental in Expedia
- RemoteRental -> RemoteRentalUnit in Expedia

I saw two options how to work around it:
1. pretend that b.com entities are the right ones. But it would introduce a lot of mess to Expedia and other apps with similar model names
2. use completely different naming for the view layer in order to decouple from channel app naming.

I chose the second approach. So the main thing I try to do here is introducing naming for models which are responsible for displaying listings for a particular channel.

## Naming

For b.com:
- Rental -> CoreListing
- RemoteAccount -> ChannelListingSection
- RemoteRental -> ChannelListing

For Expedia it is mapped like this:
- Rental -> CoreListing
- RemoteRental -> ChannelListingSection
- RemoteRentalUnit -> ChannelListing

## Notes
In order to provide interface where I specify how I collect the data I added new config option called `listings_repository_proc`. You can see an example of it in the dummy project called `ListingsRepository`.

I also found a performance issue when we search inside a particular section. We do query data which will stay intact. In order to prevent this I added setter methods which check in the beginning if it is required to query particular data based on `search_by_core_listings?` or `search_by_channel_listings?` methods.

kaizencodes commented on gemfiles/rails_5.1.gemfile.lock:
```diff
@@ -313,34 +315,34 @@ GEM
 DEPENDENCIES
   appraisal
   bookingsync_portal!
-  byebug
```
why? :D

sineed replied:
this was made unconsciously :)

kaizencodes reviewed (commented):
Good Job. some nice improvements here. I'm not 100% sure on the naming though. First of the channel listing section feels a bit off. I'd like to see the opinion of @LeoMo42 and @StoneFrog  as they worked on the initial phases of expedia and koedia and then the terms remote_rental and remote_rental_unit got locked for both apps and I think would stay for all new apps as well. I personally like the core / channel naming.

valexl reviewed (commented):
That's a great job 👍 

However I am not fully agree with approach of using this `set_core_listings` and other similar methods. Not sure we need to add arguments as a block there - check my comments please.

PS also don't forget to update Readme to explain how to use it (not sure I added it for paginated view although 😄 )

valexl commented:
And I also don't personally like that we added new entities. Before we had only rentals/remote_rentals/remote_accounts, then after adding channel builder we added remote_rental_unit, and changed meaning for remote_rentals and remote_accounts and now we added 2 new entities - core_listing and channel_listing.

sineed commented on app/controllers/bookingsync_portal/admin/rentals_controller.rb:
```diff
@@ -43,89 +40,112 @@ def action_variables
+        set_core_listings { @listings_repository.find_core_listings }
+        set_available_listings_count { @listings_repository.available_listings_count }
+
+        set_channel_listings { @listings_repository.find_channel_listings }
```
cc @kaizencodes 
All these additional wrappers came when I found one performance-related thing: when we use search bar for the core listings we make additional requests for the listing on the right side and vice versa. All these methods were aimed to fix this behavior. Initially I thought that it is required to have such endpoint because you can share the link but link is not overwritten when you do requests inside iframe. So there is no need to always return the actual picture for both lists: it is important to do that in the first request but for further ones we need to just update variables used for one of the lists.

I'm not sure if the implementation is good enough, I think it will be better to add two more controller actions (or two new controllers) in order to completely remove all these checks. But here comes other apps with their additional functionality which is added via `before_rentals_index_action_filter` and `after_rentals_index_action_filter`. So it may be not so easy and clear.  WDYT?

sineed commented on app/controllers/bookingsync_portal/admin/rentals_controller.rb:
```diff
@@ -136,8 +156,24 @@ def search_filter
-      def use_paginated_view?
-        BookingsyncPortal.use_paginated_view.call(current_account)
+      def set_core_listings
+        @action_variables.core_listings = search_by_channel_listings? ? [] : yield
```
it checks whenever client requested to search channel listings. When it happens we respond with some javascript code that replaces list on the right side with a fresh data. Nothing happens on a left side with core listings. So here I don't want to spend extra time querying data for core listings and set them as default ones (in js code we render the whole index template which expects all variables to be set and then get a specific data nodes to update the content, please refer to the `index_with_search.js.erb` file for details)

sineed commented on app/controllers/bookingsync_portal/admin/rentals_controller.rb:
```diff
@@ -43,89 +40,112 @@ def action_variables
-        @action_variables.visible_rentals = current_account.rentals.visible
-        @action_variables.remote_rentals = current_account.remote_rentals.ordered
-        @action_variables.blank_remote_accounts = generate_blank_remote_accounts
+        @listings_repository = BookingsyncPortal.listings_repository_proc.call(current_account)
```
Now I understand why he said he doesn't have a bit of time to review it :trollface:

sineed commented on bookingsync_portal.gemspec:
```diff
@@ -38,11 +38,11 @@ Gem::Specification.new do |s|
   s.add_dependency 'kaminari'
 
-  s.add_development_dependency 'byebug'
+  s.add_development_dependency 'pry'
```
sorry guys, every time I try to use byebug I recall how I'm used to `pry` and blindly replace it then :) I return it back

sineed commented on app/views/bookingsync_portal/admin/rentals/index.html.erb:
```diff
@@ -44,12 +46,12 @@
       </div>
       <div class="rentals-list-footer text-center">
-        <%= render partial: 'new_remote_account' %>
+        <%= render partial: 'new_channel_listing_section' %>
```
@kaizencodes perhaps this code may be used instead `property_creation_request` in expedia. WDYT?

kaizencodes replied:
@sineed yeah probably, you developed the ui for it so you have more insight on this.

sineed commented:
@valexl @kaizencodes 
Thanks for the comments!

> I'm not 100% sure on the naming though. channel listing section feels a bit off. I'd like to see the opinion of @LeoMo42 and @StoneFrog as they worked on the initial phases of expedia and koedia and then the terms remote_rental and remote_rental_unit were used for both apps and I think would stay for all new apps as well. I personally like the core / channel naming.

I'm not sure too. General idea was to use completely new naming for this part that should be responsible only for data rendering on this template. That is why rentals were converted to listings. For the section it is a bit tricky for me because for b.com we have `remote_accounts`, for expedia `remote_rentals` so I wasn't sure what is the best common name. Perhaps `section` is too common. Do you have any ideas here?

If I chose `remote_rental`/`remote_rental_unit` naming then we have a mess for b.com at least.

> And I also don't personally like that we added new entities. Before we had only rentals/remote_rentals/remote_accounts, then after adding channel builder we added remote_rental_unit, and changed meaning for remote_rentals and remote_accounts and now we added 2 new entities - core_listing and channel_listing.

That is the reason why I did it: I know nothing about the history :) There is not much of UI on the channel apps so I thought that adding a couple of new entities which are responsible only for data presentation will be easy and flexible to work with. Sure, you have to know about it, be ready to properly map and construct queries in the particular channel app. But in the end you forget about other entities, you work only with these new ones in every channel app and that's it.
You haven't mentioned that it is a breaking change: in order to use this PR for, say, b.com I need to apply changes and test it again.

sineed commented:
have to close because we don't want to rely on this gem in the future
